lib/common/socket.c

```diff
@@ -460,29 +458,24 @@ static void shutdown_ssl(h2o_socket_t *sock, const char *err)
         ptls_buffer_t wbuf;
         uint8_t wbuf_small[32];
         ptls_buffer_init(&wbuf, wbuf_small, sizeof(wbuf_small));
-        if ((ret = ptls_send_alert(sock->ssl->ptls, &wbuf, PTLS_ALERT_LEVEL_WARNING, PTLS_ALERT_CLOSE_NOTIFY)) != 0)
+        if (ptls_send_alert(sock->ssl->ptls, &wbuf, PTLS_ALERT_LEVEL_WARNING, PTLS_ALERT_CLOSE_NOTIFY) != 0)
             goto Close;
         write_ssl_bytes(sock, wbuf.base, wbuf.off);
         ptls_buffer_dispose(&wbuf);
-        ret = 1; /* close the socket after sending close_notify */
     } else if (sock->ssl->ossl != NULL) {
         ERR_clear_error();
-        if ((ret = SSL_shutdown(sock->ssl->ossl)) == -1)
+        if (SSL_shutdown(sock->ssl->ossl) == -1)
```

Comment:
The changes look good and I agree with your remarks on HTTP and TLS shutdown.

Perhaps unrelated to the intent of this PR, but this (and the ptls one above) direct path to `Close` looks odd to me. My understanding of `SSL_shutdown` is that it may have buffered the alert even on a -1 return.

Couldn't we always go through the `has_pending_ssl_bytes` check below? perhaps simplifying this even further?